lib/Language/Souffle/Interpreted.hs

```diff
@@ -0,0 +1,215 @@
+
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# LANGUAGE TypeFamilies, DerivingVia, InstanceSigs #-}
+module Language.Souffle.Interpreted
+  ( Program(..)
+  , Fact(..)
+  , Marshal(..)
+  , Handle
+  , ContainsFact
+  , MonadSouffle(..)
+  , SouffleM
+  , runSouffle
+  , cleanup
+  ) where
+
+import Prelude hiding (init)
+
+import Control.DeepSeq (deepseq)
+import Control.Monad.State.Strict
+import Data.IORef
+import Data.Foldable (traverse_)
+import Data.List hiding (init)
+import Data.Maybe (fromMaybe)
+import Data.Proxy
+import Data.Word
+import Language.Souffle.Class
+import Language.Souffle.Marshal
+import System.Directory
+import System.Environment
+import System.FilePath
+import System.IO.Temp
+import System.Process
+import Text.Printf
+
+
+newtype SouffleM a = SouffleM { runSouffle :: IO a }
+  deriving
+    ( Functor
+    , Applicative
+    , Monad
+    , MonadFail
+    , MonadIO
+    ) via IO
+
+
+-- | The handle for the interpreter is the path where the souffle executable
+-- can be found, and a template directory where the program is stored.
+data HandleData = HandleData
+  { soufflePath :: FilePath
+  , basePath    :: FilePath
+  , factPath    :: FilePath
+  , outputPath  :: FilePath
+  , datalogExec :: FilePath
+  , noOfThreads :: Word64
+  }
+
+newtype Handle prog = Handle (IORef HandleData)
+
+instance MonadSouffle SouffleM where
+  type Handler SouffleM = Handle
+
+  {- | Looks for a souffle interpreter.
+
+     The action will return 'Nothing' if it failed to load the Souffle program.
```

Comment:
Maybe we should improve this comment including all the erroneous cases.

Reply:
Yes, but I'm leaving docs for last 😄  I will keep this comment unresolved until it's added.

Reply:
Good point